### DISCUSSION TITLE
Artisan commands don't terminate the app on SIGINT

### DISCUSSION THREAD

**Discussion by zakini** [laravel/framework#47444] [Ideas]
### Laravel Version

10.13.5

### PHP Version

8.2.7

### Database Driver & Version

N/A

### Description

Stopping an artisan command with CTRL+C (SIGINT) appears to skip gracefully terminating the app, since no registered terminating callbacks are run when this happens.

Since we've got support for handling signals in artisan commands now, this feels like something artisan commands could do out of the box

### Steps To Reproduce

1. Clone this example repo: https://github.com/Zakini/artisan-terminating
2. Run `php artisan regular` - A "Terminating!" message will be in the logs
3. Run `php artisan long`, then stop it with CTRL+C - No new message will be in the logs
4. Run `php artisan long-term`, then stop it with CTRL+C - Another "Terminating!" message will be in the logs

**Comment by func0der** [comment]
Maybe related to this PR? https://github.com/laravel/framework/pull/47725

**Reply by macropay-solutions** [reply]
The ability to handle signals is stated in the documentation (sigterm). That should be preserved...

**Reply by func0der** [reply]
Would you happen to have a link to that part of the documentation?
The more things we have to point at and say: It is different now (or at least inconsistent), the better :D

**Reply by macropay-solutions** [reply]
https://laravel.com/docs/10.x/queues#queue-workers-and-deployment

php artisan queue:restart

This command will instruct all queue workers to gracefully exit after they finish processing their current job so that no existing jobs are lost. Since the queue workers will exit when the queue:restart command is executed, you should be running a process manager such as [Supervisor](https://laravel.com/docs/10.x/queues#supervisor-configuration) to automatically restart the queue workers.

Setting block_for to 0 will cause queue workers to block indefinitely until a job is available. This will also prevent signals such as SIGTERM from being handled until the next job has been processed.

and about supervisord : https://github.com/laravel/framework/discussions/49633

Also
https://laravel.com/docs/10.x/artisan#signal-handling

The Sigterm is also handled for workers (or was) that is Why I pasted the documentation for workers.

Also \Illuminate\Queue\Worker::listenForSignals 
```
        /**
         * Enable async signals for the process.
         *
         * @return void
         */
        protected function listenForSignals()
        {
            pcntl_async_signals(true);
    
            pcntl_signal(SIGTERM, function () {
                $this->shouldQuit = true;
            });
    
            pcntl_signal(SIGUSR2, function () {
                $this->paused = true;
            });
    
            pcntl_signal(SIGCONT, function () {
                $this->paused = false;
            });
        }
```

**Reply by func0der** [reply]
The `restart` command is for very specific use case for example a deployment and again only applies if the `worker` runs in `daemon` mode. You could also run the whole thing via `queue:listen` where signals could be completely ignored with the current setup.

The `daemon` mode of the worker is not the problem. For that mode proper signal handling is registered. Everything though is not as explained here: https://github.com/laravel/framework/pull/47725#issuecomment-1884934258

either way the PR changes the whole signal setup for EVERYTHING console, which is what @Zakini is trying to get to with his inquiry

**Reply by macropay-solutions** [reply]
Thanks for the explanation. queue:listen is used only during development time usually anyway, so it is a good news that queue:work is not affected.

**Reply by func0der** [reply]
Especially in development it is problematic, because you can not debug properly that way :D The process does not terminate after you hit the `stop` button in the IDE for example.

And `work` is affected if you do use the `--once` flag instead of the `daemon` mode.

**Reply by ghost** [reply]
Good think I never use it.
I use
queue:work;
queue:restart - to clear the cached code;
then again queue:work

**Reply by ghost** [reply]
Queue:work holds the app bootstrapped in memory just like in laravel octane which causes issues with data clogging etc. Now I'm forced to use queue:listen. Are there any issues with it in laravel 8 and 9 from your knowledge?

**Reply by func0der** [reply]
@marius-mcp You might want to open another discussion for that, because it unrelated to this issue. :)

**Reply by ghost** [reply]
@func0der thanks. https://github.com/laravel/framework/discussions/49954
